pkg/envoy/registry/services.go

```diff
@@ -0,0 +1,98 @@
+package registry
+
+import (
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/openservicemesh/osm/pkg/envoy"
+	k8s "github.com/openservicemesh/osm/pkg/kubernetes"
+	"github.com/openservicemesh/osm/pkg/service"
+)
+
+// ProxyServiceMapper knows how to map Envoy instances to services.
+type ProxyServiceMapper interface {
+	ListProxyServices(*envoy.Proxy) ([]service.MeshService, error)
+}
+
+// ExplicitProxyServiceMapper is a custom ProxyServiceMapper implementation.
+type ExplicitProxyServiceMapper func(*envoy.Proxy) ([]service.MeshService, error)
```

Comment:
Why is this needed?

Reply:
Some of the unit tests in `pkg/envoy/*ds` use a mocked `MeshCatalog` that doesn't expose any of the Kubernetes details required for `KubeProxyServiceMapper` to work.

e.g.
https://github.com/openservicemesh/osm/blob/25744ba1a1259f5863b627862850cc1333abf1c0/pkg/envoy/cds/response_test.go#L46-L48

Reply:
I left a related comment below, but I wonder why this intermediary KubeProxyServiceMapper struct is even necessary. Can't we have the call flow as `XDS -> proxy.GetServices() -> kubeController.Something()`?

Reply:
If `proxy.GetServices()` always invoked a `kubeController`, then those tests I mentioned above will break since the tests mock everything instead of creating kube pods and services (even with the fake client). Those tests could probably be rewritten to introduce a bunch of kubernetes concepts, but that would make them much harder to grok.

Since we need two ways to map proxies to services (kube and non-kube), we have an interface with two types implementing that behavior, and the proxy has a handle to an instance of one of those types. That's why we have that extra level of indirection.